app/models/user.rb

```diff
@@ -0,0 +1,21 @@
+class User < ApplicationRecord
+  before_save :downcase_email
+
+  validates :name, presence: true,
+    length: {maximum: Settings.validates.name_max_length}
+
+  validates :email, presence: true,
+    length: {maximum: Settings.validates.email_max_length},
+    format: {with: Settings.validates.email_regex},
+    uniqueness: {case_sentitive: false}
+
+  validates :password, presence: true,
+    length: {minium: Settings.validates.password_min_length}
+
+  has_secure_password
```

Comment:
dòng này có tác dụng gì e ?

Reply:
Dòng 12, 13: validate password có độ dài tối thiểu là pass_min_length
Dòng 15: Sử dụng method này để băm mật khẩu, đảm bảo sự an toàn cho dữ liệu đặc biệt là mật khẩu, kẻ tấn công sẽ không thể đăng nhập vào trang web ngay cả khi chúng quản lý để lấy được bản sao của cơ sở dữ liệu.

Reply:
chưa đủ e

Reply:
Dòng 12, 13: Ngoài ra sử dụng presence: true để kiểm tra password cần tồn tại (không được rỗng).
Dòng 15: 
- Lưu thuộc tính password_digest được băm vào database.
- Sử dụng method authenticate để trả về người dùng password có đúng hay không.
- Sử dụng hàm băm bcrypt (được thêm vào Gemfile)